src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -64,7 +63,7 @@ public sealed partial class TokenCache : ITokenCacheInternal
                 await requestParams.AuthorityManager.GetInstanceDiscoveryEntryAsync().ConfigureAwait(false);
 
             #region Create Cache Objects
-            if (!string.IsNullOrEmpty(response.AccessToken))
+            if (!string.IsNullOrEmpty(response.AccessToken) && !requestParams.ShouldNotCacheAccessTokens)
```

Comment:
instead of relying on `requestParams` to tell you this (which it shouldn't, because it doesn't deal with caching), better to add the logic here: 

`
bool shouldCache = (msalTokenResponse.TokenSource == Broker && msalTokenResponse.TokenType == POP) 
`